Subprocess-Worker: Faulthandler und Output filtern und sofort loggen

MartinGauk opened the pull request:
* Bei debug Log-Level sollen die Ausgaben des Workers im stderr sofort ausgegeben werden.
* Andernfalls werden die stderr Ausgaben als INFO nun geloggt.
* ASCII Control Characters werden aus Sicherheitsgründen ersetzt. Besonders die ANSI Escape Codes sollen nicht im Terminal ankommen.
* Der Faulthandler von Python wird aktiviert und dem Prozess wird ein SIGABRT geschickt, wenn er wegen einer Zeitüberschreitung gestoppt werden soll.
https://docs.python.org/3/library/faulthandler.html

Mit diesen Maßnahmen soll das Debugging verbessert werden.

MartinGauk committed:
feat: log worker stderr immediately on debug level and filter ASCII control chars

MartinGauk committed:
feat: faulthandler in subprocess worker and print traceback on time limit exceptions

janbritz commented on questionpy_server/worker/impl/subprocess.py:
```diff
@@ -159,6 +193,22 @@ def _get_observation_tasks(self) -> Sequence[asyncio.Task]:
+
+                # We give it a short time to dump the traceback. Usually, SIGABRT should terminate the process
+                # right after Python's signal handler returns.
+                try:
+                    await asyncio.wait_for(self._proc.wait(), 0.2)
+                except TimeoutError:
+                    log.warning("Worker %s did not terminate after SIGABRT", self.name)
+                else:
+                    return
```
Ich finde das etwas lesbarer:
```suggestion
                try:
                    await asyncio.wait_for(self._proc.wait(), 0.2)
                    return
                except TimeoutError:
                    log.warning("Worker %s did not terminate after SIGABRT", self.name)
```

MartinGauk replied:
Das hatte ich ursprünglich so geschrieben und fände ich auch lesbarer, aber ruff hat mich auf https://docs.astral.sh/ruff/rules/try-consider-else/ hingewiesen.

janbritz reviewed (approved):
Beim Testen ist mir aufgefallen, dass beim Beenden des Servers mit `CTRL + C` pro laufendem `SubprocessWorker` eine `KeyboardInterrupt` Exception geloggt wird und die Worker anscheindend nicht gracefully gestoppt werden. Vielleicht könnten wir deshalb beim `bootstrap`en des Workers das Signal ignorieren: `signal.signal(signal.SIGINT, signal.SIG_IGN)`.

<details>
<summary>Logs</summary>
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output: Traceback (most recent call last):
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:   File "<frozen runpy>", line 198, in _run_module_as_main
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:   File "<frozen runpy>", line 88, in _run_code
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:   File "/home/jan/work/qpy-dev/questionpy/server/questionpy_server/worker/runtime/__main__.py", line 7, in <module>
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:     subprocess_runtime_main()
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:   File "/home/jan/work/qpy-dev/questionpy/server/questionpy_server/worker/runtime/subprocess_main.py", line 37, in subprocess_runtime_main
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:     manager.loop()
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:   File "/home/jan/work/qpy-dev/questionpy/server/questionpy_server/worker/runtime/manager.py", line 104, in loop
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:     msg = self._connection.receive_message()
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:   File "/home/jan/work/qpy-dev/questionpy/server/questionpy_server/worker/runtime/connection.py", line 44, in receive_message
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:     header_bytes = self.stream_in.read(messages_header_struct.size)
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output:                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
DEBUG:questionpy_server.worker.impl.subprocess:Worker 1dd8c39235-1 output: KeyboardInterrupt
</details>

MartinGauk committed:
fix: start subprocess worker in its own process group and session

When pressing CTRL+C, the terminal should not send SIGINT to our workers
as this will make them stop ungracefully.

See discussion in PR #151.

MartinGauk commented:
> Beim Testen ist mir aufgefallen, dass beim Beenden des Servers mit `CTRL + C` pro laufendem `SubprocessWorker` eine `KeyboardInterrupt` Exception geloggt wird und die Worker anscheindend nicht gracefully gestoppt werden.

Das Terminal sendet `SIGINT` auch an unsere Subprocesses. Ich denke, wir sollten die Worker in eine eigene Process Group stecken, damit die das Signal nicht mehr erhalten. Wenn ich den Prozess mit [`start_new_session=True`](https://docs.python.org/3/library/subprocess.html#popen-constructor) starte, tritt das Problem nicht auf und die Worker werden normal durch den questionpy-server beendet.

Ich hatte noch überlegt, ob man das auch als Feature betrachten kann, weil so sichergestellt wird, dass die Worker beendet werden, selbst wenn der Server wegen eines Bugs das Beenden nicht selber durchführt, aber wenn wir Container-Worker haben, dann sind die ohnehin in ihrer eigenen Process Group.
Anders als im Terminal stoppt/killt Systemd die Prozesse einer Unit anhand ihrer cgroup. Dann müssten auch die Subprocess-Worker gekillt werden. Bei Docker/Podman gibt es die `--cgroup-parent` Option. Möglicherweise können wir die Container-Worker in eine Unter-cgroup der Systemd-Unit stecken. Aber da müssten wir uns nochmal genauer informieren, ob das so gut ist, wenn den Containern via Systemd ein SIGKILL gesendet wird.
